docs/source/more_types.rst

```diff
@@ -933,3 +933,210 @@ and non-required keys, such as ``Movie`` above, will only be compatible with
+above use cases can't be solved this way. For such situations, one might want
+to use ``typing.Final``.
+
+Definition syntax
+-----------------
+
+The ``typing.Final`` type qualifier indicates that a given name or attribute
+should never be re-assigned, re-defined, nor overridden. It can be used in
+one of these forms:
+
+* The simplest one is ``ID: Final = 1``. Note that unlike gor generic classes
+  this is *not* the same as ``Final[Any]``. Here mypy will infer type ``int``.
+
+* An explicit type ``ID: Final[float] = 1`` can be used as in any
+  normal assignment.
+
+* In stub files one can omit the right hand side and just write
+  ``ID: Final[float]``.
+
+* Finally, one can define ``self.id: Final = 1`` (also with a type argument),
+  but this is allowed *only* in ``__init__`` methods.
+
+Definition rules
+----------------
+
+The are two rules that should be always followed when defining a final name:
+
+* There can be *at most one* final declaration per module or class for
+  a given attribute:
+
+  .. code-block:: python
+
+     ID: Final = 1
+     ID: Final = 2  # Error!
+
+     class SomeCls:
+         id: Final = 1
+         def __init__(self, x: int) -> None:
+             self.id: Final = x  # Error!
+
+  Note that mypy has a single namespace for a class. So there can't be two
+  class-level and instance-level constants with the same name.
+
+* There must be *exactly one* assignment to a final attribute:
+
+  .. code-block:: python
+
+     ID = 1
+     ID: Final = 2  # Error!
+
+     class SomeCls:
+         ID = 1
+         ID: Final = 2  # Error!
+
+.. note::
+   Conditional final declarations and final declarations within loops are
+   not supported.
+
+Using final attributes
+----------------------
+
+As a result of a final declaration mypy strives to provide the
+two following guarantees:
+
+* A final attribute can't be re-assigned (or otherwise re-defined), both
+  internally and externally:
+
+  .. code-block:: python
+
+     # file mod.py
+     from typing import Final
+
+     ID: Final = 1
+
+     # file main.py
+     from typing import Final
+
+     import mod
+     mod.ID = 2  # Error, can't assign to constant.
+
+     class SomeCls:
+         ID: Final = 1
+
+         def meth(self) -> None:
+             self.ID = 2  # Error, can't assign to final attribute
+
+     class DerivedCls(SomeCls):
+         ...
+
+     DerivedCls.ID = 2  # Error!
+     obj: DerivedCls
+     obj.ID = 2  # Error!
+
+* A final attribute can't be overridden by a subclass (even with another
+  explicit final declaration). Note however, that final attributes can
+  override normal attributes. This also applies to multiple inheritance:
+
+  .. code-block:: python
+
+     class Base:
+         ID = 0
+
+     class One(Base):
+         ID: Final = 1  # OK
+
+     class Other(Base):
+         ID: Final = 2  # OK
+
+     class Combo(One, Other):  # Error, cannot override final attribute.
+         pass
+
```

Comment:
Maybe also mention that this doesn't turn the object assigned to a final attribute immutable. For example, this would be fine:

```
x: Final = ['a', 'b']
x.append('x')  # Ok
```

Using a read-only ABC or protocol can be used to work around this sometimes (or an immutable concrete class, of course):

```
x: Final[Sequence[str]] = ['a', 'b']
x.append('x')  # Error
y: Final = ('a', 'b')  # Also an option
```